### ISSUE TITLE
Graphql Query with nested objects by implementing a custom data provider without Doctrine

### ISSUE THREAD

**Issue by safwanghamrawi** [api-platform/core#2833]
I am using Symfony 4 with api platform 2.4. I am not using Doctrine and so neither Entities. I have two classes (Not Entities) Facet and Option. There is a relation between them so I define an attribute options in Facet class which is a collection of Option (like OneToMany in Doctrine). I have my custom data provider because I am getting those data from an external api. I build my collection of facets with options in this custom data provider.
Here is my classes

```
<?php

namespace App\Model;

use ApiPlatform\Core\Annotation\ApiProperty;
use ApiPlatform\Core\Annotation\ApiResource;
use Doctrine\Common\Collections\ArrayCollection;
use \Doctrine\Common\Collections\Collection;

/**
 * @ApiResource()
 * Class Facet
 *
 * @package App\Model
 */
class Facet
{
    /**
     * @ApiProperty(identifier=true)
     * @var int
     */
    private $id;

    /**
     * @var string
     */
    private $name;

    /**
     * @var \App\Model\Option[]
     */
    private $options;

    public function __construct()
    {
        $this->options = new ArrayCollection();
    }

    /**
     * @return int
     */
    public function getId(): int
    {
        return $this->id;
    }

    /**
     * @param int $id
     *
     * @return Facet
     */
    public function setId(int $id): self
    {
        $this->id = $id;

        return $this;
    }

    /**
     * @return string
     */
    public function getName(): string
    {
        return $this->name;
    }

    /**
     * @param string $name
     *
     * @return Facet
     */
    public function setName(string $name): self
    {
        $this->name = $name;

        return $this;
    }


    /**
     * @return Collection|Option[]
     */
    public function getOptions(): Collection
    {
        return $this->options;
    }

    public function addOption(Option $option): self
    {
        if (!$this->options->contains($option)) {
            $this->options[] = $option;
            $option->setFacet($this);
        }

        return $this;
    }

    public function removeOption(Option $option): self
    {
        if ($this->options->contains($option)) {
            $this->options->removeElement($option);
            // set the owning side to null (unless already changed)
            if ($option->getFacet() === $this) {
                $option->setFacet(null);
            }
        }

        return $this;
    }
}

<?php

namespace App\Model;

use ApiPlatform\Core\Annotation\ApiProperty;
use ApiPlatform\Core\Annotation\ApiResource;

/**
 * Class Option
 *
 * @ApiResource()
 * @package App\Model
 */
class Option
{
    /**
     * @ApiProperty(identifier=true)
     * @var int
     */
    private $id;

    /**
     * @var string
     */
    private $value;

    /**
     * @var int
     */
    private $count;

    /**
     * @var Facet
     */
    private $facet;

    /**
     * @return int
     */
    public function getId(): int
    {
        return $this->id;
    }

    /**
     * @param int $id
     *
     * @return Option
     */
    public function setId(int $id): self
    {
        $this->id = $id;

        return $this;
    }

    /**
     * @return string
     */
    public function getValue(): string
    {
        return $this->value;
    }

    /**
     * @param string $value
     *
     * @return Option
     */
    public function setValue(string $value): self
    {
        $this->value = $value;

        return $this;
    }

    /**
     * @return int
     */
    public function getCount(): int
    {
        return $this->count;
    }

    /**
     * @param int $count
     *
     * @return Option
     */
    public function setCount(int $count): self
    {
        $this->count = $count;

        return $this;
    }

    /**
     * @return Facet
     */
    public function getFacet(): Facet
    {
        return $this->facet;
    }

    /**
     * @param Facet $facet
     *
     * @return Option
     */
    public function setFacet(Facet $facet = null): self
    {
        $this->facet = $facet;

        return $this;
    }
}

<?php

namespace App\DataProvider;

use ApiPlatform\Core\DataProvider\CollectionDataProviderInterface;
use ApiPlatform\Core\DataProvider\RestrictedDataProviderInterface;
use App\Model\Facet;
use App\Model\Option;
use Doctrine\Common\Collections\ArrayCollection;

/**
 * Class FacetDataProvider
 *
 * @package App\DataProvider
 */
final class FacetDataProvider implements CollectionDataProviderInterface, RestrictedDataProviderInterface
{
    public function supports(string $resourceClass, string $operationName = null, array $context = []): bool
    {
        return Facet::class === $resourceClass;
    }

    public function getCollection(string $resourceClass, string $operationName = null)
    {
        $k = 0;
        $collection = new ArrayCollection();
        for ($i = 0; $i < 2; $i++) {
            $facet = new Facet();
            $facet
                ->setId(++$i)
                ->setName("facet [$i]");
            for ($j = 0; $j < 2; $j++) {
                $option = new Option();
                $option
                    ->setId($k++)
                    ->setValue("option [$i] [$j]")
                    ->setCount(rand(0, 10))
                ;
                $facet->addOption($option);
            }

            $collection->add($facet);
        }

        return $collection;
    }
}
```
In Graphql Doc, the options is iterable but it should be a ConnectionOption

```
id: ID!
_id: Int!
name: String!
options: Iterable!
```
so I can't execute graphql query on nested object, I have en error: field "options" must not have a selection since type iterable! has no subfields

```
{
  facets {
    edges {
      node {
        id
        name
        options {
          edges {
            node {
              id
              value
              count
            }
          }
        }
      }
    }
  }
}
```

For information I tried to use doctrine ORM annotation to workaround this problem and to have ConnectionOption instead of iterable and it works but when I am executing this graphql query, I have always an empty array for options (it seems like they are not serialized or even not existing)
```
{
  "data": {
    "facets": {
      "edges": [
        {
          "node": {
            "id": "/api/facets/1",
            "name": "facet [1]",
            "options": []
          }
        }
      ]
    }
  }
}
```

I create those classes as Entity to show the expected results for Graphql Doc

```
id: ID!
_id: Int!
name: String!
options(first: Int after: String): OptionConnection
```

and when execute the graphql query and select iat
```
{
  facets {
    edges {
      node {
        id
        name
        options {
          edges {
            node {
              id
              count
              value
            }
          }
        }
      }
    }
  }
}
```

This is the response what I want without doctrine

```
{
  "data": {
    "facets": {
      "edges": [
        {
          "node": {
            "id": "/api/facets/1",
            "name": "facet [1]",
            "options": {
              "edges": [
                {
                  "node": {
                    "id": "/api/options/1",
                    "count": 0,
                    "value": "option [1] [0]"
                  }
                },
                {
                  "node": {
                    "id": "/api/options/2",
                    "count": 0,
                    "value": "option [1] [1]"
                  }
                }
              ]
            }
          }
        }
      ]
    }
  }
}
```

To sum up, I want to use custom dataprovider with php classes and nested object without doctrine

Thanks for everyone

**Comment by alanpoulain** [comment]
Should be solved by https://github.com/api-platform/core/pull/3327.
